### PR TITLE
Adding documentation to the IEx.Info protocol

### LINKED ISSUE #11262: IEx.Info protocol documentation?
When calling `i/1` in IEx, general information about the term passed to it is shown.
This is implemented as the `IEx.Info` protocol.

I am interested in implementing this protocol for my own types, to improve the description "This is a struct. Structs are maps with a __struct__ key." as well as potentially add some other definitions to the list of definitions.

However, IEx.Info is not currently documented at all.

- Should we add documentation to it?
- Or is this protocol intended to be 'private' to Elixir/IEx? If so, should we add some indication that this is the case somewhere?

### PR DESCRIPTION
**Pull request by Qqwy**
Fixes #11262 

This is my initial stab at adding this documentation.
It's based on my own interpretation of the code of the existing `IEx.Info`-implementations as well as `IEx.Helpers.i/1`.

Please let me know in what ways this documentation can be improved. I expect that parts are too specific, and in other parts important information might be missing.

:slightly_smiling_face:

### CODE REVIEW

File: lib/iex/lib/iex/info.ex [hunk 1]
```diff
@@ -1,6 +1,45 @@
...
+  the `"Data type"`, `"Description"` and `"Reference modules"` sections.
+  """
+
+  @doc """
+  Returns information for the given term.
+
+  Information should be returned as a list of `info_name`-`info` pairs,
+
+  `info_name` should be short.
+  It is OK for `info` to contain newlines.
```
Comment (josevalim) [inline lib/iex/lib/iex/info.ex:20]:
```suggestion
  Information should be returned as a list of `info_name`-`info` tuples,
  where both values are strings. `info_name` should be short. `info` can
  be arbitrarily long and contain newlines.
```

File: lib/iex/lib/iex/info.ex [hunk 2]
```diff
@@ -1,6 +1,45 @@
...
+     defining the data type.
+  - `"Description"`: One or a few sentences describing what the data type represents.
+  - `"Reference modules`: One or a few comma-separated module names that focus
+    on working with this datatype.
+
+  Other recommended sections are:
+
+  - `"Raw representation`: showing another way of writing the passed `term`.
+    This is mostly relevant for data-structures whose `String.Chars`-implementations
+    make use of sigils or other syntactic sugar.
```
Comment (josevalim) [inline lib/iex/lib/iex/info.ex:41]:
```suggestion
    * `"Data type"`: Name of the data type. Usually the name of the module
       defining the data type.
    * `"Description"`: One or a few sentences describing what the data type represents.
    * `"Reference modules`: One or a few comma-separated module names that focus
      on working with this datatype.

  Other recommended sections are:

    * `"Raw representation`: showing another way of writing the passed `term`.
      This is mostly relevant for data-structures whose `String.Chars`-implementations
      make use of sigils or other syntactic sugar.
```

Just standardizing with other markdown in the codebase. :)

**Review by josevalim — COMMENTED** [review]
Two tiny suggestions and we can ship it!

File: lib/iex/lib/iex/info.ex (hunk 1 above)
Comment (Qqwy) [reply lib/iex/lib/iex/info.ex:20]: I'm for this change, but note that the spec:

```
  @spec info(term()) :: [{info_name :: String.Chars.t(), info :: String.t()}]
```

as well as [iex/helpers.ex:503](https://github.com/elixir-lang/elixir/blob/master/lib/iex/lib/iex/helpers.ex#L503) seem to hint that there are cases where `info_name` is not necessarily a string. 
Do we want to alter the spec and potentially remove the call to `to_string`?

Comment (josevalim) [reply lib/iex/lib/iex/info.ex:20]: I updated the suggestion. Let me know what you think.

**Comment by Qqwy** [comment]
Suggestions incorporated :+1: ! Ready to ship when you are :ship: .

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**